### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Alternating

### PR DESCRIPTION
**Pull request by Komyyy**
---

- [x] depends on: leanprover-community/mathlib#18766
* [`linear_algebra.alternating`@`ce11c3c2a285bbe6937e26d9792fda4e51f3fe1a`..`284fdd2962e67d2932fa3a79ce19fcf92d38e228`](https://leanprover-community.github.io/mathlib-port-status/file/linear_algebra/alternating?range=ce11c3c2a285bbe6937e26d9792fda4e51f3fe1a..284fdd2962e67d2932fa3a79ce19fcf92d38e228)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 1]
```diff
@@ -0,0 +1,1262 @@
...
+section Coercions
+
+-- Porting note: Replaced CoeFun with FunLike instance
+instance funLike : FunLike (AlternatingMap R M N ι) (ι → M) (fun _ => N) where
+  coe f := f.toFun
+  coe_injective' := fun f g h ↦ by
+    rcases f with ⟨⟨_, _, _⟩, _⟩
+    rcases g with ⟨⟨_, _, _⟩, _⟩
+    cases h
+    rfl
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean:107]: I think it would be reasonable to wait for https://github.com/leanprover-community/mathlib/pull/18766

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Alternating.lean:107]: (I don't have much time this weekend; feel free to push to that PR to get it working, if you want things to move faster)

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib#18766
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by Parcly-Taxel** [comment]
leanprover-community/mathlib#18766 should land before the next mathport run, in a few hours from this post. You should be able to change the SHAs and add in the new stuff and then we'll be done with this file.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib#18766~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by eric-wieser** [comment]
> [leanprover-community/mathlib#18766](https://github.com/leanprover-community/mathlib/pull/18766) should land before the next mathport run, in a few hours from this post. You should be able to change the SHAs and add in the new stuff and then we'll be done with this file.


I think this isn't true, you'll probably need to wait 27 hours

**Comment by Komyyy** [comment]
https://github.com/leanprover-community/mathlib3port/blob/master/Mathbin/LinearAlgebra/Alternating.lean
It seems that the PR was in the time.

**Comment by eric-wieser** [comment]
Oh, excellent!

**Review by Parcly-Taxel — APPROVED** [review]

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 2]
```diff
@@ -103,8 +103,11 @@ instance funLike : FunLike (AlternatingMap R M N ι) (ι → M) (fun _ => N) whe
   coe_injective' := fun f g h ↦ by
     rcases f with ⟨⟨_, _, _⟩, _⟩
     rcases g with ⟨⟨_, _, _⟩, _⟩
-    cases h
-    rfl
+    congr
+
+-- Porting note: The shortcut instance is removed in Lean4.
+--  instance coeFun : CoeFun (AlternatingMap R M N ι) fun _ => (ι → M) → N :=
+--   ⟨FunLike.coe⟩
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean:109]: Is there any reason to remove it?

Comment (Komyyy) [reply Mathlib/LinearAlgebra/Alternating.lean:109]: According to the porting note on `Mathlib.Algebra.Hom.Ring`, the shortcut instance is unhelpful in Lean4:
```lean
-- Porting note:
-- These helper instances are unhelpful in Lean 4, so omitting:
-- /-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
-- directly.
-- -/
-- instance : CoeFun (α →+* β) fun _ => α → β :=
--   ⟨RingHom.toFun⟩
```

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Alternating.lean:109]: They're only actively bad if they mention `toFun` like that one does. This one uses `FunLike.coe` so is harmless.

Perhaps worth checking on Zulip whether we view these as phones.

Comment (Komyyy) [reply Mathlib/LinearAlgebra/Alternating.lean:109]: OK, I add this instance.

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 3]
```diff
@@ -90,23 +92,21 @@ variable (v : ι → M) (v' : ι → M')
...
-  coe_injective' f g h := by
-    cases f
-    cases g
+instance funLike : FunLike (AlternatingMap R M N ι) (ι → M) (fun _ => N) where
+  coe f := f.toFun
+  coe_injective' := fun f g h ↦ by
+    rcases f with ⟨⟨_, _, _⟩, _⟩
+    rcases g with ⟨⟨_, _, _⟩, _⟩
     congr
-#align alternating_map.fun_like AlternatingMap.funLike
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean]: Why was this deleted?

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 4]
```diff
@@ -90,23 +92,21 @@ variable (v : ι → M) (v' : ι → M')
...
 
--- shortcut instance
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean]: Any reason for deleting this comment?

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 5]
```diff
@@ -115,24 +115,29 @@ theorem toFun_eq_coe : f.toFun = f :=
...
 #align alternating_map.to_fun_eq_coe AlternatingMap.toFun_eq_coe
 
+-- Porting note: New theorem.
 @[simp]
-theorem coe_mk (f : (ι → M) → N) (h₁ h₂ h₃) : ⇑(⟨f, h₁, h₂, h₃⟩ : AlternatingMap R M N ι) = f :=
+theorem coe_mk (f : MultilinearMap R (fun _ : ι => M) N) (h) :
+    ⇑(⟨f, h⟩ : AlternatingMap R M N ι) = f :=
+  rfl
+
+theorem coe_mks (f : (ι → M) → N) (h₁ h₂ h₃) : ⇑(⟨⟨f, h₁, h₂⟩, h₃⟩ : AlternatingMap R M N ι) = f :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean:124]: I think the thing to do would be to delete this lemma, and align `alternating_map.coe_mk` with `alternating_map.coe_mk\_x`

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 6]
```diff
@@ -115,24 +115,29 @@ theorem toFun_eq_coe : f.toFun = f :=
   rfl
 #align alternating_map.to_fun_eq_coe AlternatingMap.toFun_eq_coe
 
+-- Porting note: New theorem.
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean:118]:
```suggestion
-- Porting note: Changed statement to reflect new `mk` signature.
```

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 7]
```diff
@@ -143,30 +148,30 @@ theorem ext {f f' : AlternatingMap R M N ι} (H : ∀ x, f x = f' x) : f = f' :=
...
-@[simp]
-theorem toMultilinearMap_eq_coe : f.toMultilinearMap = f :=
-  rfl
-#align alternating_map.to_multilinear_map_eq_coe AlternatingMap.toMultilinearMap_eq_coe
+#noalign alternating_map.to_multilinear_map_eq_coe
 
 @[simp]
 theorem coe_multilinearMap_mk (f : (ι → M) → N) (h₁ h₂ h₃) :
-    ((⟨f, h₁, h₂, h₃⟩ : AlternatingMap R M N ι) : MultilinearMap R (fun i : ι => M) N) =
+    ((⟨⟨f, h₁, h₂⟩, h₃⟩ : AlternatingMap R M N ι) : MultilinearMap R (fun _ : ι => M) N) =
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean:174]: This should be changed to
```suggestion
    ((⟨f, h₃⟩ : AlternatingMap R M N ι) : MultilinearMap R (fun _ : ι => M) N) =
```
along with a porting note referring to `mk` changing, and a subscript `x` on the align.

File: Mathlib/LinearAlgebra/Alternating.lean [hunk 8]
```diff
@@ -491,25 +501,25 @@ theorem compLinearMap_assoc (f : AlternatingMap R M N ι) (g₁ : M₂ →ₗ[R]
...
 #align alternating_map.add_comp_linear_map AlternatingMap.add_compLinearMap
 
 @[simp]
 theorem compLinearMap_zero [Nonempty ι] (f : AlternatingMap R M N ι) :
     f.compLinearMap (0 : M₂ →ₗ[R] M) = 0 := by
   ext
-  simp_rw [comp_linear_map_apply, LinearMap.zero_apply, ← Pi.zero_def, map_zero, zero_apply]
+  -- Porting note: Was `simp_rw [.., ← Pi.zero_def, map_zero, zero_apply]`.
+  simp_rw [compLinearMap_apply, LinearMap.zero_apply, ← Pi.zero_def, zero_apply]
+  exact map_zero f
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Alternating.lean:522]: Did adding back the shortcut instance make the old proof work again?

Comment (Parcly-Taxel) [reply Mathlib/LinearAlgebra/Alternating.lean:522]: It doesn't; `exact map_zero f` is still required.

File: Mathlib/LinearAlgebra/Alternating.lean (hunk 5 above)
Comment (Parcly-Taxel) [reply Mathlib/LinearAlgebra/Alternating.lean]: I've just pushed a better fix for this and `coe_multilinearMap_mk`. Noting that the mathport output has
```lean
theorem coe_mk (f : (ι → M) → N) (h₁ h₂ h₃) : ⇑(⟨f, h₁, h₂, h₃⟩ : AlternatingMap R M N ι) = f :=
  rfl

theorem coe_multilinearMap_mk (f : (ι → M) → N) (h₁ h₂ h₃) :
    ((⟨f, h₁, h₂, h₃⟩ : AlternatingMap R M N ι) : MultilinearMap R (fun i : ι => M) N) =
      ⟨f, @h₁, @h₂⟩ :=
  rfl
```
and that these don't work, but putting angle brackets in the right places (as already existed in this PR) fixes everything, it only makes sense to just add the brackets and leave porting notes explaining the new `mk` structure.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**